### ISSUE TITLE
Inactivity causes freeze

### ISSUE THREAD

**Issue by wlfbck** [shinyhut/vernacular-vnc#5]
I'm currently encountering a problem where i'm unfortunately not sure if it's related to vernacular, the VNC server or a combination of both.

What happens:
I can connect to the server just fine, everything looks great, mouse and keyboard events are transmitted and working fine. Then, if i do nothing for 30 seconds, the screen "freezes".
 I logged the timings of when i get a frames.

This is the pattern for the first 30 seconds:

>25.03.2020 11:14:36.447: got frame
>25.03.2020 11:14:37.433: got frame
>25.03.2020 11:14:38.502: got frame
>25.03.2020 11:14:39.423: got frame
>25.03.2020 11:14:40.636: got frame
>25.03.2020 11:14:41.492: got frame
>25.03.2020 11:14:42.213: got frame
>25.03.2020 11:14:42.517: got frame
>25.03.2020 11:14:43.622: got frame
>25.03.2020 11:14:44.479: got frame
>25.03.2020 11:14:45.546: got frame
>25.03.2020 11:14:46.507: got frame
>25.03.2020 11:14:47.459: got frame

This is the pattern after 30 seconds:
>25.03.2020 11:19:50.290: got frame
>25.03.2020 11:19:54.352: got frame
>25.03.2020 11:19:56.379: got frame
>25.03.2020 11:19:58.417: got frame
>25.03.2020 11:20:00.442: got frame
>25.03.2020 11:20:02.481: got frame
>25.03.2020 11:20:04.505: got frame
>25.03.2020 11:20:08.568: got frame
>25.03.2020 11:20:14.667: got frame
>25.03.2020 11:20:18.730: got frame
>25.03.2020 11:20:22.792: got frame
>25.03.2020 11:20:26.856: got frame
>25.03.2020 11:20:28.881: got frame
>25.03.2020 11:20:32.946: got frame
>25.03.2020 11:20:39.48: got frame
>25.03.2020 11:20:43.112: got frame
>25.03.2020 11:20:47.162: got frame
>25.03.2020 11:20:51.226: got frame
>25.03.2020 11:20:59.354: got frame
>25.03.2020 11:21:01.390: got frame
>25.03.2020 11:21:05.454: got frame
>25.03.2020 11:21:09.516: got frame

Connecting with a different VNC client to the server instantly fixes the problem and unfreezes the vernacular client. Sometimes it also randomly fixes itself, but not by any action that i could logically relate with it.

Our VNC server is quite old (x11vnc: 0.9.13 lastmod: 2011-08-10 ), but for our normal needs we used UltraVNC (latest version) and never ran into this issue.

Any help is greatly appreciated :)

**Comment by wlfbck** [comment]
Additional info: When it unfreezes itself, i can see all the mouse events i did while it was frozen being processed.

**Comment by wlfbck** [comment]
Here's the output from connecting with vernacular:
```
25/03/2020 13:25:26 Got connection from client 192.168.23.162
25/03/2020 13:25:26   other clients:
25/03/2020 13:25:26 Normal socket connection
25/03/2020 13:25:26 Disabled X server key autorepeat.
25/03/2020 13:25:26   to force back on run: 'xset r on' (3 times)
25/03/2020 13:25:26 incr accepted_client=16 for 192.168.23.162:3714  sock=12
25/03/2020 13:25:26 Client Protocol Version 3.8
25/03/2020 13:25:26 Protocol version sent 3.8, using 3.8
25/03/2020 13:25:26 rfbProcessClientSecurityType: executing handler for type 1
25/03/2020 13:25:26 rfbProcessClientSecurityType: returning securityResult for client rfb version >= 3.8
25/03/2020 13:25:27 Pixel format for client 192.168.23.162:
25/03/2020 13:25:27   16 bpp, depth 16, big endian
25/03/2020 13:25:27   true colour: max r 31 g 31 b 63, shift r 11 g 0 b 5
25/03/2020 13:25:27 Enabling NewFBSize protocol extension for client 192.168.23.162
25/03/2020 13:25:27 Using hextile encoding for client 192.168.23.162
25/03/2020 13:25:27 client useCopyRect: 192.168.23.162 -1
25/03/2020 13:25:27 client_set_net: 192.168.23.162  0.0000
25/03/2020 13:25:27 created   xdamage object: 0x400060
```

And then also connecting with UltraVNC:
```
25/03/2020 13:35:15 Got connection from client 192.168.23.162
25/03/2020 13:35:15   other clients:
25/03/2020 13:35:15      192.168.23.162
25/03/2020 13:35:15 Normal socket connection
25/03/2020 13:35:15 incr accepted_client=17 for 192.168.23.162:4096  sock=13
25/03/2020 13:35:16 Client Protocol Version 3.8
25/03/2020 13:35:16 Protocol version sent 3.8, using 3.8
25/03/2020 13:35:16 rfbProcessClientSecurityType: executing handler for type 1
25/03/2020 13:35:16 rfbProcessClientSecurityType: returning securityResult for client rfb version >= 3.8
25/03/2020 13:35:16 Pixel format for client 192.168.23.162:
25/03/2020 13:35:16   16 bpp, depth 16, little endian
25/03/2020 13:35:16   true colour: max r 31 g 63 b 31, shift r 11 g 5 b 0
25/03/2020 13:35:16 no translation needed
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0x0000000A)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type zlibhex
25/03/2020 13:35:17 Using compression level 6 for client 192.168.23.162
25/03/2020 13:35:17 Enabling full-color cursor updates for client 192.168.23.162
25/03/2020 13:35:17 Enabling cursor position updates for client 192.168.23.162
25/03/2020 13:35:17 Using image quality level 6 for client 192.168.23.162
25/03/2020 13:35:17 Using JPEG subsampling 0, Q79 for client 192.168.23.162
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type xorEnable
25/03/2020 13:35:17 Enabling LastRect protocol extension for client 192.168.23.162
25/03/2020 13:35:17 Enabling NewFBSize protocol extension for client 192.168.23.162
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xFFFF8000)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xFFFF8001)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xFFFF8004)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xFFFF8002)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xFFFF8003)
25/03/2020 13:35:17 rfbProcessClientNormalMessage: ignoring unsupported encoding type Enc(0xC0A1E5CE)
25/03/2020 13:35:17 Using hextile encoding for client 192.168.23.162
25/03/2020 13:35:17 client useCopyRect: 192.168.23.162 -1
25/03/2020 13:35:17 client_set_net: 192.168.23.162  0.0000
25/03/2020 13:35:17 client useCopyRect: 192.168.23.162 -1
25/03/2020 13:35:17 active keyboard: turning X autorepeat off.
```

Notice how ultra also sets a bunch of more options, not sure if it is related to that.
Also being connected with UltraVNC at the same time does not prevent the problem.

**Comment by wlfbck** [comment]
Alright, if turn down the framerate option to 1 (10 wasnt enough), it is not freezing anymore. I still see the slowdown of received frames when inactive, but it is not causing freezes anymore.

Several interesting things:
- UltraVNC is definitely using a higher framerate, and is also getting frames with roughly that updaterate.
- UltraVNC also seems to request and get "small" updates on mouse movement (it's extremely smooth, but definitely not emulated client side, you can see how it lags behind always a little bit).

I suspect that atleast one of the options that UltraVNC is setting is causing this "better" behavior, but unsure which. My guess would be [LastRect](https://vncdotool.readthedocs.io/en/0.8.0/rfbproto.html#lastrect-pseudo-encoding). The stuff about the cursor can be ignored, i was able to turn that off in UltraVNC and it didn't affect anything.

**Comment by shinyhut** [comment]
Thanks for the detailed bug report @wlfbck, I'll try to reproduce it and work out what's going on.

**Comment by shinyhut** [comment]
@wlfbck I've reproduced the issue exactly as you describe using x11vnc 0.9.13 and I think I have a fix. Can you try this experimental build from the [bugfix/5-inactivity-causes-screen-freeze](https://github.com/shinyhut/vernacular-vnc/tree/bugfix/5-inactivity-causes-screen-freeze) branch:
[vernacular-1.8-SNAPSHOT.jar.zip](https://github.com/shinyhut/vernacular-vnc/files/4397722/vernacular-1.8-SNAPSHOT.jar.zip)?

The problem was x11vnc didn't like the aggressive way vernacular polls for incremental framebuffer updates at its target framerate even if the server had no updates to send (i.e. the display hasn't changed). This change ensures we don't request framebuffer updates faster than the server sends them.

**Comment by wlfbck** [comment]
Working great, thank you very much for the swift response and fix! 😄

**Comment by shinyhut** [comment]
Excellent! That's been merged and released as [v1.8.0](https://github.com/shinyhut/vernacular-vnc/releases/tag/v1.8.0)
